components/pd_client/src/client_v2.rs

```diff
@@ -0,0 +1,1358 @@
+            }
+            Err(e) => {
+                warn!("failed to reconnect pd"; "err" => ?e);
+            }
+            _ => {}
+        }
+    }
+}
+
+impl RpcClient {
+    pub fn new(
+        cfg: &Config,
+        shared_env: Option<Arc<Environment>>,
+        security_mgr: Arc<SecurityManager>,
+    ) -> Result<RpcClient> {
+        let env = shared_env.unwrap_or_else(|| {
+            Arc::new(
+                EnvBuilder::new()
+                    .cq_count(CQ_COUNT)
+                    .name_prefix(thd_name!(CLIENT_PREFIX))
+                    .build(),
+            )
+        });
+
+        let raw_client = CachedRawClient::new(cfg.clone(), env.clone(), security_mgr);
+
+        let lame_client = PdClientStub::new(Channel::lame(env, "0.0.0.0:0"));
+        lame_client.spawn(reconnect_loop(raw_client.clone(), cfg.clone()));
+
+        Ok(Self {
+            raw_client,
+            feature_gate: Default::default(),
+        })
+    }
+
+    pub fn set_on_reconnect<F: Fn() + Sync + Send + 'static>(&self, f: F) {
+        let lame_client =
+            PdClientStub::new(Channel::lame(self.raw_client.env().clone(), "0.0.0.0:0"));
+        let mut rx = self.raw_client.clone().on_reconnect_rx;
+        lame_client.spawn(async move {
+            loop {
+                if rx.recv().await.is_err() {
+                    break;
+                } else {
+                    f();
+                }
+            }
+        });
+    }
+
+    #[cfg(feature = "testexport")]
+    pub fn feature_gate(&self) -> &FeatureGate {
+        &self.feature_gate
+    }
+
+    #[cfg(feature = "testexport")]
+    pub fn get_leader(&mut self) -> pdpb::Member {
+        block_on(self.raw_client.wait_for_ready()).unwrap();
+        self.raw_client.leader()
+    }
+
+    #[cfg(feature = "testexport")]
+    pub fn reconnect(&mut self) -> Result<bool> {
+        block_on(self.raw_client.reconnect())
+    }
+
+    #[cfg(feature = "testexport")]
+    pub fn cluster_id(&self) -> u64 {
+        self.raw_client.cluster_id()
+    }
+}
+
+pub trait PdClient {
+    type ResponseChannel<R: Debug>: Stream<Item = Result<R>>;
+
+    fn create_region_heartbeat_stream(
+        &mut self,
+        wake_policy: mpsc::WakePolicy,
+    ) -> Result<(
+        mpsc::Sender<RegionHeartbeatRequest>,
+        Self::ResponseChannel<RegionHeartbeatResponse>,
+    )>;
+
+    fn create_report_region_buckets_stream(
+        &mut self,
+        wake_policy: mpsc::WakePolicy,
+    ) -> Result<mpsc::Sender<ReportBucketsRequest>>;
+
+    fn create_tso_stream(
+        &mut self,
+        wake_policy: mpsc::WakePolicy,
+    ) -> Result<(mpsc::Sender<TsoRequest>, Self::ResponseChannel<TsoResponse>)>;
+
+    fn load_global_config(&mut self, list: Vec<String>) -> PdFuture<HashMap<String, String>>;
+
+    fn watch_global_config(
+        &mut self,
+    ) -> Result<grpcio::ClientSStreamReceiver<pdpb::WatchGlobalConfigResponse>>;
+
+    fn get_cluster_id(&mut self) -> Result<u64>;
```

Comment:
Why need to be a `Result`? And isn't it just the same as L476?

Reply:
`wait_for_ready` can fail. The L476 version will simply panic. I'm removing the latter now.